### PR TITLE
[pt1][tensor] caffe2::DeviceType -> at::DeviceType

### PR DESCRIPTION
**Pull request by jerryzh168**
[pt1][tensor] caffe2::DeviceType -> at::DeviceType

Previously we use DeviceType in caffe2.proto directly, but it's an `enum` and have implicit conversion to int, which does not have type safety, e.g. we have to explicitly check for a device type is valid in event.h:
```
template <int d>
struct EventCreateFunctionRegisterer {
  explicit EventCreateFunctionRegisterer(EventCreateFunction f) {
    static_assert(d < MaxDeviceTypes, "");
    Event::event_creator_[d] = f;
  }
};
```
at::DeviceType is an `enum class`, and it does not have implicit conversion to int, and provides better type safety guarantees. In this diff we have done the following refactor(taking CPU as an example):

    1. caffe2::DeviceType → caffe2::DeviceTypeProto
    2. caffe2::CPU → caffe2::PROTO_CPU
    3. caffe2::DeviceType = at::DeviceType
    4. caffe2::CPU = at::DeviceType::CPU

codemod -d caffe2/caffe2 --extensions h,cc,cpp 'device_type\(\), ' 'device_type(), PROTO_'
+ some manual changes

In short, after this diff, in c++, caffe2::CPU refers to the at::DeviceType::CPU and the old proto caffe2::CPU will be caffe2::PROTO_CPU.
In python side, we have a temporary workaround that alias `caffe2_pb2.CPU = caffe2_pb2.PROOT_CPU` to make the change easier to review and this will be removed later.

Differential Revision: D9545704

### CODE REVIEW

File: caffe2/core/tensor.h [hunk 1]
```diff
@@ -222,17 +222,18 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
     if (size() > 0) {
       if (storage_.dtype().copy()) {
         CAFFE_ENFORCE(
-            GetDeviceType() == CPU,
+            GetDeviceType() == DeviceType::CPU,
```
Comment (ezyang) [inline caffe2/core/tensor.h]: You didn't end up adding an `constexpr auto CPU = DeviceType::CPU` somewhere global in the Caffe2 namespace? (It's OK either way, just a little surprised!)

File: caffe2/python/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+from caffe2.proto import caffe2_pb2
+# TODO: refactor & remove the following alias
```
Comment (ezyang) [inline caffe2/python/__init__.py]: Nice to have a little comment here saying what this is all about, given that we had so much discussion about how to do this :)

**Review by ezyang — APPROVED** [review]

File: caffe2/core/tensor.h (hunk 1 above)
Comment (jerryzh168) [reply caffe2/core/tensor.h]: oh we have that, this should be a change I did before I define that.
